File: data-structures/linked_list.py
# a linked list is a sequences of nodes where each node store its own data and a link to the next node 
# no preallocated space is needed ,  insertion is easier
# Functions were going to be implementing 
# insert at begining /
# insert at end /
# insert at /
# insert_values /
# insert_after_value /
# remove_by_value /
# remove_at 
# print_list /
# get_list_length /


class Node:
    def __init__(self, head=None , next=None):
        self.head = head 
        self.next = next 

class LinkedList:
    def __init__(self , head=None):
        self.head = head 
    #inserting at the start of the list
    def insert_at_begining(self , data):
        node= Node(data ,self.head)
        self.head=node
    
    #inserting at the end of the list 
    def insert_at_end(self , data):
        itr = self.head 
        if self.head is None:
            node = Node(data , self.head)
            self.head =node 
            return 
        #iterating over each item in our list 
        while itr:
            #checking if the next item is none 
            # if its none the we'll know we're at the end 
            if itr.next is None :
                #inserting a new node at the end  of the list 
                node = Node(data , None)
                itr.next = node
                break 
            itr = itr.next
            
    # inserting a fresh set of values into our list 
    def insert_values(self , data_list):
        #clearing out the head
        self.head = None 
        #looping through the items in the list and appending them to the linked list 
        for data_item in data_list :
            self.insert_at_end(data_item)
    
    def insert_at(self , index , data):
        if index < 0 or index > self.get_list_length() -1:
            raise Exception("invalid index")
        # if the index is at 0 , just insert at the begining 
        if index == 0:
            self.insert_at_begining(data)
            return 
        # else 
        count = 0
        itr = self.head
        while itr :
            if count == index - 1:
                node = Node(data , itr.next)
                itr.next= node
                break
            count +=1
            itr = itr.next
        # removing item by a value 
    def remove_by_value(self, data_name):
        itr = self.head 
        while itr :
            # when the next data_name is found replace it with what comes after it 
            if itr.next.head == data_name:
                itr.next =itr.next.next
                break 
            if itr.next is None:
                break 
            itr = itr.next
            
    def remove_at(self , index):
        if index < 0 or index > self.get_list_length() -1 :
            raise Exception("invalid index")
        if index == 0:
            itr = self.head 
            self.head = itr.next
            return
        
        count = 0 
        itr = self.head 
        while itr:
            if count == index -1 :
                # node = Node(None , itr.next)
                itr.next = itr.next.next
                break
            itr = itr.next
            count +=1
            
    #inserting after a value 
    def insert_after_value(self , data_after , data_to_insert):
        itr = self.head
        while itr :
            # when the element is found 
            if itr.head == data_after:
                #create a node 
                node = Node(data_to_insert , itr.next)
                #replace the former node with the new node
                itr.next = node 
                break
            if itr.next is None :
                break
            itr = itr.next
            
    #printing out the contents of the list     
    def print_list(self):
        itr = self.head 
        if self.head is None :
            print("List is currently empty")
            return 
        list=''
        while itr :
            list += f"{itr.head} --> "
            itr = itr.next
        print(list)
    
    def get_list_length(self):
        itr = self.head
        count = 0 
        #when the next iteration item is None break the loop 
        while itr:
            count +=1 
            itr = itr.next
        return count 
    #inserting after a specified value 

            
        
li = LinkedList()
li.insert_values([i*2 for i in range(1,10)])
li.print_list()
li.insert_after_value(6 , 10)
li.remove_by_value(6)
li.insert_at_begining(20)
li.insert_at_end(20)
li.print_list()
li.remove_at(2)
li.remove_at(3)
li.remove_at(0)
li.remove_at(0)
li.print_list()

File: data-structures/javascript/linked_list.js
class Node {
  constructor(head = null, next = null) {
    this.head = head;
    this.next = next;
  }
}

class LinkedList {
  constructor() {
    this.head = null;
  }

  /* addding to add_to_begining */
  add_to_begining(new_item) {
    let node = new Node(new_item, this.head);
    this.head = node;
  }
  /* adding to end */
  add_to_end(new_item) {
    if (this.head === null) {
      this.add_to_begining(new_item);
      return;
    }
    let itr = this.head;
    while (itr) {
      if (itr.next === null) {
        itr.next = new Node(new_item, null);
        break;
      }
      itr = itr.next;
    }
  }

  /* inserting at */
  insert_at(index, new_item) {
    if (index-1 > this.length()-1 || index < 0) {
      throw new Error("Invalid index");
    }
    if (index === 0) {
      this.add_to_begining(new_item);
      return;
    }
    let itr = this.head;
    let count = 0;
    while (itr) {
      if (count+1 == index) {
        itr.next = new Node(new_item, itr.next);
        break;
      }
      count++;
      itr = itr.next;
    }
  }

  /* replace at */
  delete_at(index) {
    if (index-1 > this.length()-1 || index < 0) {
      throw new Error("Invalid index");
    }
    if (index === 0) {
      this.head = this.head.next;
      return;
    }
    let itr = this.head;
    let count = 0;
    while (itr) {
      if (count+1 == index) {
        itr.next = itr.next.next;
        break;
      }
      count++;
      itr = itr.next;
    }
  }

  /* checking if empty */
  isEmpty() {
    return this.head? false: true;
  }
  length() {
    if (this.head === null) {
      return 0;
    }
    let itr = this.head;
    let count = 0;
    while (itr) {
      count++;
      itr = itr.next;
    }
    return count;
  }

  /* insering many values */
  insert_values(item_list) {
    this.head = null;
    item_list.forEach(item => this.add_to_end(item));
  }
  /* printing node */
  print_list() {
    let itr = this.head;
    let listr = "";
    while (itr) {
      listr += ` ${itr.head} --> `;
      itr = itr.next;
    }
    console.log(listr);
  }

}

const li = new LinkedList();

li.add_to_end(10);
li.add_to_end(20);
li.add_to_begining(30);
li.add_to_end(130);
li.print_list();
li.insert_at(4, "name");
li.print_list();
li.delete_at(2);
li.print_list();
li.insert_values(["2", "4", 6 , 10 , 20]);
li.print_list();
li.delete_at(3);
li.print_list();
console.log(li.length());